### PR TITLE
[TT-5396] Add ability to select basic authentication hashing function

### PR DESCRIPTION
**Pull request by sedkis**
Add ability to select basic authentication hashing function, instead of always "bcrypt"

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit:  
Triggered by: pull_request (@sedkis)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2378651716)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: 158d351c27721ed6f9db0b94d0d1e61faf0bcf40 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547198156)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: 9b977cf2ee5e1bab960029ea402a8ea1cdab1a88 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547331633)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: 4f93228d8c796273b6180d73029df790781e1c8c 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547675236)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: a789164aa45fba1420c83ff58932bc04d3b2fb29 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547797733)

**Comment by Tyk-ITS** [comment]
**API tests result: failure** :no_entry_sign: 
Branch used: refs/pull/4070/merge
Commit: 75e71f73b301783f90f1e6bcfacb23bbc19b7541 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547953586)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: c86928b62861b58f28d4d827681bc1f2754c6ee6 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2547999789)

**Comment by buger** [comment]
Overall looks cool! 

The only tests I miss is to ensure that backward compatibility still works. E.g. you have key which has old bcrypt password, you have a key with a new algorithm, and both of them work in parallel.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: 2aaec392fbe6a2c695a68087cb3d2ea5b856ad71 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2549756625)

**Comment by titpetric** [comment]
Side effect: after upgrading gateway with the new default field, `basic_auth_hash_key_function` will default to `bcrypt`; If an operator has set `hash_key_function=sha256` before the upgrade, this would change the hash function that's used on requests from sha256 to bcrypt.

If we want to keep the current behavior, `basic_auth_hash_key_function` can be empty/unset, and we can use `hash_key_function` for basic auth requests, as we did so far. The current behavior is to use "bcrypt", but given the side effect, I would suggest to amend this.

@buger if there's a test case which we still can do, some hints how to put it together are appreciated.

**Comment by titpetric** [comment]
You are correct, i am incorrect that this was configurable. Leaves my last concern resolved :)

**Review by furkansenharputlu — APPROVED** [review]

File: cli/linter/schema.json [hunk 1]
```diff
@@ -448,13 +448,25 @@
         "type": "string",
         "enum": [
           "",
+          "bcrypt",
```
Comment (buger) [inline cli/linter/schema.json:451]: bcypt not needed here, e.g. we should not has key IDs using it.

Comment (buger) [reply cli/linter/schema.json:451]: Also based on the code, we have not modified this part too, so just needs be removed from here ,and from the code comment below

**Review by buger — APPROVED** [review]

File: cli/linter/schema.json (hunk 1 above)
Comment (titpetric) [reply cli/linter/schema.json]: Removed! :+1:

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4070/merge
Commit: 8a1ba755e38e9da229717ae6b0fce81334ea29f6 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2574679864)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 4aeddcbed92690763e58feb44eef8df7038f2d83 [TT-5396] Add ability to select basic authentication hashing function (#4070)

* Add ability to select basic authentication hashing function, instead of always "bcrypt"

* Fix linter schema to add basic_auth_hash_key_function

* storage: sanitize input hash algo

* gateway: add TestBasicAuthHashKeyFunc for code coverage

* storage: add TestHashAlgo for coverage

* gateway: use available hash algos for basic auth

* gateway: add BasicAuthKeyIsValid.checkPassword for unit tests

* Fix existing tests in auth_manager

* gateway: Extend TestBasicAuthHashKeyFunc, refactor for coverage

* Update config schema specs and comment with inputs

* Remove bcrypt from hash_key_function values

Co-authored-by: Tit Petric <tit.petric@monotek.net>
Triggered by: push (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2574756817)

**Pull request merged**